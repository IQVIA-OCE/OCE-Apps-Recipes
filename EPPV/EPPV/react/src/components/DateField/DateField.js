import React, { useRef, useState } from 'react';
import { View } from 'react-native';
import { DateTime } from 'luxon';
import { TextInput, DateTimePicker, useTheme } from 'apollo-react-native';

export const DateField = ({ value, onChange, label, hasError, helperText, disabled }) => {
  const [isVisibleDatePicker, setVisibleDatePicker] = useState(false);
  const dateInput = useRef();
  const theme = useTheme();

  const formatDate = (date) => {
    return DateTime.fromISO(date)
      .toLocaleString(DateTime.DATE_SHORT)
      .replace(/\./g, '/');
  };

  return (
    <View>
      <TextInput
        ref={dateInput}
        label={label}
        value={value ? formatDate(value) : ''}
        icon={'calendar-blank'}
        iconColor={theme.colors.text}
        onFocus={() => setVisibleDatePicker(true)}
        onIconPress={() => setVisibleDatePicker(true)}
        fullWidth
        required
        asteriskPosition={'after'}
        error={hasError}
        helperText={helperText}
        disabled={disabled}
        testID={'datePickerInput'}
      />
      <DateTimePicker
        visible={isVisibleDatePicker}
        anchor={dateInput}
        date={value ? new Date(value) : new Date()}
        mode={'datetime'}
        onBackdropPress={() => setVisibleDatePicker(false)}
        actions={{
          done: {
            onPress: (_, d) => {
              onChange(d.toISOString());
              setVisibleDatePicker(false);
            },
          },
        }}
        testID={'datePicker'}
      />
    </View>
  );
};
