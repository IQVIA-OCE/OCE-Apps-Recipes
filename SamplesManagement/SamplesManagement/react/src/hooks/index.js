export { useFetcher } from './useFetcher';
export { useBoolean } from './useBoolean';
export { useHandleData } from './useHandleData';
export { useBanner } from './useBanner';
