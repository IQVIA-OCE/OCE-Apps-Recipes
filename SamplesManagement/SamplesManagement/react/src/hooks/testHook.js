import renderer from 'react-test-renderer';
import React from 'react';

const TestHook = ({ callback }) => {
  callback();
  return null;
};

export const testHook = callback =>
  renderer.create(<TestHook callback={callback} />);

const TestRenderHook = ({ callback }) => {
  return callback();
};

export const testRenderHook = callback =>
  renderer.create(<TestRenderHook callback={callback} />);
