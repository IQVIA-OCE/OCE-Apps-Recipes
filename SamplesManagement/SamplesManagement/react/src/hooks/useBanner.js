import { useEffect, useState } from 'react';

export const useBanner = () => {
  const [banner, setBanner] = useState({
    variant: '',
    message: '',
    visible: false,
    icon: '',
  });

  useEffect(() => {
    const timer = setTimeout(
      () => setBanner(prev => ({ ...prev, visible: false })),
      4000
    );

    return () => clearTimeout(timer);
  }, [banner.visible]);

  return [banner, setBanner];
};
