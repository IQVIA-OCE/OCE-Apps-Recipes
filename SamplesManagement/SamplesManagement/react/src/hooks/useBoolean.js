import { useState } from 'react';

export const useBoolean = initialValue => {
  const [value, setValue] = useState(initialValue);

  const toggle = () => setValue(v => !v);
  const setTrue = () => setValue(true);
  const setFalse = () => setValue(false);

  return [value, { toggle, setTrue, setFalse }];
};
