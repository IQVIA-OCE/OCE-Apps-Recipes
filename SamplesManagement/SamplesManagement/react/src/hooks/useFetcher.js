import { useCallback, useEffect, useState } from 'react';

export const useFetcher = (fetch, normalizer) => {
  const [value, setValue] = useState({
    loading: true,
    error: '',
    data: null,
    metadata: null,
  });

  const handleFetch = async () => {
    setValue(prevState => ({
      ...prevState,
      loading: true,
    }));

    try {
      let [data, metadata] = await fetch();
      if (typeof normalizer === 'function') {
        data = normalizer(data);
      }
      setValue({
        loading: false,
        error: '',
        data,
        metadata,
      });
    } catch (error) {
      setValue({
        loading: false,
        error: error,
        data: null,
        metadata: null,
      });
    }
  };

  useEffect(() => {
    handleFetch();
  }, []);

  return [value, { handleFetch, setValue }];
};
