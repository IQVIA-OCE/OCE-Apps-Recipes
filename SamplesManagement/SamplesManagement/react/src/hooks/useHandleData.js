import { Text, View } from 'react-native';
import { ActivityIndicator, Colors } from 'apollo-react-native';
import React from 'react';

export const useHandleData = ({ loading, error, data }) => {
  return (fn) => {
    if (loading)
      return (
        <ActivityIndicator
          animating={true}
          color={Colors.blue700}
          style={{ paddingVertical: 10 }}
        />
      );
    if (error)
      return (
        <View>
          <Text>{error && error.message || error}</Text>
        </View>
      );

    return fn(data);
  };
};
