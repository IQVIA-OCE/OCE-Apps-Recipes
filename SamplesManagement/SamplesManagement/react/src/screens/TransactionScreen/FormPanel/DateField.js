import React, { useRef } from 'react';
import { View, StyleSheet, Text } from 'react-native';
import moment from 'moment';
import { TextInput, DateTimePicker } from 'apollo-react-native';
import { useBoolean } from '../../../hooks';

const DateField = ({ value, onChange, label, style, required, errors, touched, hasError, helperText }) => {
  const [isVisible, visibleActions] = useBoolean(false);
  const input = useRef();
  return (
    <View style={[styles.container, style]}>
      <TextInput
        ref={input}
        label={label}
        value={value ? moment(value).format('MMM DD, YYYY') : ''}
        icon="calendar-blank"
        iconColor="#000"
        onFocus={visibleActions.setTrue}
        onIconPress={visibleActions.setTrue}
        fullWidth
        required={required}
        error={hasError}
        helperText={helperText}
      />
      <DateTimePicker
        visible={isVisible}
        anchor={input}
        date={new Date()}
        mode={'date'}
        onBackdropPress={visibleActions.setFalse}
        actions={{
          done: {
            onPress: (_, d) => {
              onChange(d);
              visibleActions.setFalse();
            },
          },
        }}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {},
});

export default DateField;
