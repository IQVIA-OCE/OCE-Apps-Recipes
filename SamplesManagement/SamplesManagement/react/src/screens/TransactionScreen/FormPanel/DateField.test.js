import React from 'react';
import DateField from './DateField';
import renderer, { act } from 'react-test-renderer';
import { DateTimePicker } from 'apollo-react-native';
import { useBoolean } from '../../../hooks';

jest.mock('../../../hooks');

describe('DateField', () => {
  beforeEach(() => {
    useBoolean.mockReturnValue([
      false,
      { toggle: jest.fn(), setFalse: jest.fn(), setTrue: jest.fn() },
    ]);
  });
  it('should render component', async () => {
    const promise = Promise.resolve();
    const onChange = jest.fn();
    let tree;

    act(() => {
      tree = renderer.create(<DateField onChange={onChange} />);
    })

    act( () => {
      tree.root.findByType(DateTimePicker).props.actions.done.onPress();
    })
    await act(() => promise);
    expect(onChange).toHaveBeenCalled();
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
