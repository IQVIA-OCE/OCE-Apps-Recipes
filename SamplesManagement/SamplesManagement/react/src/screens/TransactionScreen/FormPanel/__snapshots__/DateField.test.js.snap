// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateField should render component 1`] = `
<View
  style={
    Array [
      Object {},
      undefined,
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "width": "100%",
        },
        undefined,
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "borderColor": "#D9D9D9",
            "borderRadius": 4,
            "borderWidth": 1,
          },
          Object {},
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 40,
            "minHeight": 40,
            "overflow": "hidden",
            "paddingBottom": 0,
            "paddingTop": 0,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          maxResizableWidth={750}
          minHeight={60}
          minWidth={0}
          onBlur={[Function]}
          onFocus={[Function]}
          paddingLeft={8}
          paddingRight={40}
          placeholderTextColor="rgba(0, 0, 0, 0.54)"
          rejectResponderTermination={true}
          selectionColor="#0768FD"
          style={
            Array [
              Object {
                "margin": 0,
                "textAlign": "left",
                "zIndex": 5,
              },
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "height": 40,
                "minHeight": 0,
                "paddingLeft": 8,
                "paddingRight": 40,
                "paddingVertical": 5,
                "textAlignVertical": "center",
                "width": "100%",
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "position": "absolute",
              "right": 10,
              "top": "50%",
              "transform": Array [
                Object {
                  "translateY": -12.5,
                },
              ],
              "zIndex": 5,
            }
          }
        >
          <Text
            accessibilityElementsHidden={true}
            allowFontScaling={false}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontSize": 25,
                },
                Array [
                  Object {
                    "transform": Array [
                      Object {
                        "scaleX": 1,
                      },
                    ],
                  },
                  Object {
                    "backgroundColor": "transparent",
                  },
                ],
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            ïƒ®
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View>
    <View />
  </View>
</View>
`;
